### PR TITLE
Migrate to arkade for CLIs and dependent apps

### PR DESCRIPTION
**Pull request by alexellis**
Signed-off-by: Alex Ellis (OpenFaaS Ltd) <alexellis2@gmail.com>

## Description

Migrate to arkade for CLIs and dependent apps

Migrates to arkade for checking environment, installing CLIs, 
and for apps (helm charts). This reduces ongoing maintenance and 
moves any options required into arkade.

As such, some names may change to defaults such as the name used
for Minio and SealedSecrets.

## How Has This Been Tested?

Tested without TLS or auth, a function built and was deployed on my computer with k3d.

### CODE REVIEW

**Comment by alexellis** [comment]
@Waterdrips / @wilsonianb PTAL

**Comment by alexellis** [comment]
Minio is incorrectly configured due to the app creating its own secrets and the apply command wanting to create its own ones too. A build and deployment works, but there are no logs stored.

**Comment by alexellis** [comment]
I've fixed the issue for Minio and restored the early namespace pre-creation step.

**Review by wilsonianb — COMMENTED** [review]
`ofc-bootstrap apply` is failing for me with:
```
2020/12/15 03:01:44 Creating secrets for functions to consume
2020/12/15 03:01:44 Installing openfaas
Error: plan failed after 18.192070s, error: unable to install openfaas: non-zero exit-code: Using Kubeconfig: /etc/rancher/k3s/k3s.yaml
Using kubeconfig: 
Node architecture: ""
Client: "x86_64", "Linux"
Using Kubeconfig: 
[Warning] unable to create secret basic-auth, may already exist: error: Missing or incomplete configuration info.  Please point to an existing, complete config file:

  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.
"openfaas" has been added to your repositories
VALUES values.yaml
Command: /root/.arkade/bin/helm [upgrade --install openfaas openfaas/openfaas --namespace openfaas --wait --values /tmp/charts/openfaas/values.yaml --set openfaasImagePullPolicy=IfNotPresent --set basic_auth=true --set functionNamespace=openfaas-fn --set ingress.enabled=false --set gateway.writeTimeout=15m --set gateway.upstreamTimeout=14m55s --set queueWorker.ackWait=15m --set faasnetes.httpProbe=true --set basicAuthPlugin.replicas=1 --set queueWorker.replicas=2 --set queueWorker.maxInflight=1 --set serviceType=NodePort --set gateway.readTimeout=15m --set faasnetes.readTimeout=5m --set faasnetes.writeTimeout=5m --set clusterRole=false --set gateway.directFunctions=true --set operator.create=false --set ingressOperator.create=false --set faasIdler.dryRun=false --set faasnetes.imagePullPolicy=IfNotPresent --set gateway.replicas=2 --set gateway.scaleFromZero=true]
 2020/12/15 03:01:44 User dir established as: /root/.arkade/

Error: Kubernetes cluster unreachable
Error: exit code 1, stderr: Error: Kubernetes cluster unreachable
```
on a server with k3s where I've run `export KUBECONFIG=/etc/rancher/k3s/k3s.yaml` and can access the cluster with `kubectl`

File: cmd/apply.go [hunk 1]
```diff
@@ -484,97 +465,171 @@ func createFunctionsAuth() error {
...
 
 	return nil
 }
 
-func installMinio() error {
-	log.Println("Creating Minio")
+func getS3Credentials() (string, string, error) {
+	var accessKey, secretKey string
+
+	args := []string{"get", "secret", "-n", "openfaas", "s3-access-key", "-o jsonpath='{.data.s3-access-key}'"}
```
Comment (Waterdrips) [inline cmd/apply.go:580]: These secrets live in the openfaas-fn namespace

Comment (alexellis) [reply cmd/apply.go:580]: I'll change it, thanks for catching the problem.

File: scripts/create-namespaces.sh [hunk 2]
```diff
@@ -1,11 +1 @@
 #!/bin/bash
-
```
[the remark this answers is not in the document]
Comment (alexellis) [reply scripts/create-namespaces.sh:2]: 👍

**Comment by alexellis** [comment]
@wilsonianb 
Are you logging into your k3s node and running ofc-bootstrap on it directly?

arkade (and therefore ofc-bootstrap) shouldn't be run as root. It's not supposed to be run on the computer hosting your Kubernetes cluster, but from your laptop or client machine. k3sup for instance fetches the kubeconfig file from a remote host after installing k3s, changing the permissions in the process.

I tested with k3d, so that's a similar client/server configuration.

File: cmd/registry_login.go [hunk 3]
```diff
@@ -42,19 +42,33 @@ func generateRegistryAuthFile(command *cobra.Command, _ []string) error {
...
-			return generateFile(username, password, server)
+			generateErr = generateFile(username, password, server)
 		}
 	}
+
+	if generateErr != nil {
+		return generateErr
+	}
+
+	fmt.Printf("\nWrote ./credentials/docker.json..OK\n")
```
[the remark this answers is not in the document]
Comment (alexellis) [reply cmd/registry_login.go:69]: That's a typo, thanks for catching.

**Comment by alexellis** [comment]
Tested with OAuth, Minio and @Waterdrips covered cert-manager

**Comment by wilsonianb** [comment]
>Are you logging into your k3s node and running ofc-bootstrap on it directly?

Yep, as root...

Trying again locally, I ran into a similar error, which I believe is due to https://github.com/alexellis/arkade/pull/258/files#r544694671

`ofc-bootstrap apply` ran successfully after I moved my kubeconfig to `$HOME/.kube/config`

**Pull request merged**